File: JavaScript/0001.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // Create a dictionary to store the number and its index
    let numToIndex = new Map();
    
    // Iterate through the list
    for (let i = 0; i < nums.length; i++) {
        let num = nums[i];
        // Calculate the complement that would sum to the target
        let complement = target - num;
        
        // If the complement is found in the dictionary, return the indices
        if (numToIndex.has(complement)) {
            return [numToIndex.get(complement), i];
        }
        
        // Otherwise, add the number and its index to the dictionary
        numToIndex.set(num, i);
    }
    
    // Return an empty array if no solution is found
    return [];
};

File: Ruby/0001.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    # Create a hash to store the number and its index
    num_to_index = {}
    
    # Iterate through the list
    nums.each_with_index do |num, i|
      # Calculate the complement that would sum to the target
      complement = target - num
      
      # If the complement is found in the hash, return the indices
      if num_to_index.has_key?(complement)
        return [num_to_index[complement], i]
      end
      
      # Otherwise, add the number and its index to the hash
      num_to_index[num] = i
    end
    
    # In case no solution is found, return an empty array (though the problem assumes one solution exists)
    []
end